Executor v0.3

shrijitsingh99 opened the pull request:
Changes:
~1. `instance_of_base` accepts multiple executor types~
~2. Best fit executor~
3. Document

shrijitsingh99 committed:
Bump version to 0.3

shrijitsingh99 committed:
Update instance_of_base to accept multiple supported executor types

Causes API breakage since the trait now expects executor type then the executor types to check against.
Earlier it was the reverse

shrijitsingh99 committed:
Make shapes const

kunaltyagi commented on executor/include/executor/trait/is_instance_of_base.hpp:
```diff
@@ -14,18 +14,21 @@
 
-template <template <typename...> class Type, typename Executor>
+template <typename Executor, template <typename...> class... Type>
+using is_instance_of_base = executor::disjunction<is_instance_of_base_impl<Executor, Type>...>;
```
The Executor coming before Type... might require a small change in name of `is_instance_of_base` into `is_instance_of_all` or `is_instance_of_any` (depending on conjunction/disjunction)

shrijitsingh99 replied:
True. I think we should also finalize the name for `is_instance_of_base` to whatever is more descriptive.
@SergioRAgostinho

kunaltyagi replied:
Please take a decision @shrijitsingh99, let him comment on the decision :smile:

shrijitsingh99 replied:
Renamed to `InstanceOfBase`

shrijitsingh99 committed:
Fix bug in index_type

shrijitsingh99 committed:
Add max_threads in OMP executor

kunaltyagi commented on executor/include/executor/default/omp_executor.hpp:
```diff
@@ -36,6 +36,16 @@ struct omp_executor {
     int idx;
   };
 
+  std::size_t max_threads;
```
Use size_type please

shrijitsingh99 replied:
`size_type` of which container? I didn't quite follow you.

kunaltyagi replied:
Oops, my mistake. `shape_type`

kunaltyagi commented on executor/include/executor/default/omp_executor.hpp:
```diff
@@ -36,6 +36,16 @@ struct omp_executor {
+  std::size_t max_threads;
+
+  omp_executor() {
+    this->max_threads = omp_get_max_threads();
```
Why not `omp_executor(): max_threads(omp_get_max_threads()) {}`?

shrijitsingh99 replied:
`omp_get_max_threads` would not be available on non-omp devices.
0e0d6cd adds the OMP availabilty checks

kunaltyagi replied:
Ideally, you should delegate to the other ctor in this.

The other ctor should use initializer list instead of in-body assignment

kunaltyagi replied:
```cpp
omp_executor(): omp_executor(0) {}
```
Keep macros in one place

shrijitsingh99 committed:
Add OpenMP availability checks

kunaltyagi commented on executor/include/executor/default/omp_executor.hpp:
```diff
@@ -36,13 +36,27 @@ struct omp_executor {
+  omp_executor(std::size_t max_threads): max_threads(max_threads) {
+#ifdef _OPENMP
+    if (!this->max_threads) this->max_threads = omp_get_max_threads();
+#endif
```
`ifndef _OPENMP`, should `max_threads` should be forced 0?

shrijitsingh99 replied:
I don't think so, since it is just to specify the maximum number of threads, so its the responsibility of the user to ensure OMP is available if it isn't he will anyway not be able to call using the executor.

kunaltyagi commented on executor/include/executor/default/omp_executor.hpp:
```diff
@@ -54,11 +68,14 @@ struct omp_executor {
 
   template <typename F>
-  void bulk_execute(F&& f, shape_type n) const {
+  void bulk_execute(F&& f, const shape_type n) const {
```
If OpenMP is unavailable, then should this function even exist? :thinking:

shrijitsingh99 replied:
I added static asserts to his. It will be easier to understand the error this way than just making the function non-existent case OMP is unavailable.

SergioRAgostinho replied:
The assumption is that it will only trigger compilation failures if the user explicitly tries to use it right? Is best_fit capable of triggering the instantiation in case OMP is not available?

shrijitsingh99 replied:
In 87580f6 I added this.
So if OMP is not available the best fit executor will just skip over OMP executor and it will not trigger a compilation failed and will also never be selected.

kunaltyagi commented on executor/include/executor/trait/is_instance_of_base.hpp:
```diff
@@ -14,18 +14,21 @@
+template <typename Executor, template <typename...> class... Type>
 using instance_of_base =
-    std::enable_if_t<is_instance_of_base_v<Type, Executor>, int>;
+    std::enable_if_t<is_instance_of_base_v<Executor, Type...>, int>;
```
For constexpr things, include static_assert tests in the code

shrijitsingh99 replied:
Will start adding these tests with the documentation.
This brings up the question, should the tests for executors also be migrated to PCL?

SergioRAgostinho replied:
Yes.

shrijitsingh99 committed:
Add additional type traits

shrijitsingh99 committed:
Add best fit functionality

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -51,8 +50,18 @@ void mmul(const Executor ex, const MatrixXd& a, const MatrixXd& b,
+          bool custom_priority) {
+  auto mul = [&](auto& exec) { mmul(exec, a, b, c); };
+
+  if (custom_priority)
```
Note to self:
1. what is custom priority?
2. What is descending priority?

Custom decay policy or the library default decay policy?

SergioRAgostinho commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,111 @@
+  static bool check(Executor& exec) {
+    return true;
+  }
+};
```
All of them return true. What's the intended use here?

shrijitsingh99 replied:
So currently we don't have any runtime checks but since we have decided some runtime checks they will go here and return true/false based on the runtime conditions.
https://github.com/shrijitsingh99/unified-executors/issues/22

SergioRAgostinho commented on executor/include/executor/default/cuda_executor.hpp:
```diff
@@ -69,7 +69,7 @@ struct cuda_executor {
   // some reason
   template <typename F>
-  void bulk_execute(F& f, shape_type shape) const {
+  void bulk_execute(F& f, const shape_type shape) const {
```
shouldn't shape be passed by const reference here? Sometimes they're integrals, but sometimes they're not.

shrijitsingh99 committed:
Add static asserts for when user tries to use an unavailable executor

shrijitsingh99 committed:
Add checks for availability of executor

SergioRAgostinho commented on executor/include/executor/trait/common_traits.hpp:
```diff
@@ -38,4 +55,84 @@ struct equality_comparable<
+
+template <typename Executor, template <typename...> class... Type>
+using is_instance_of_base =
+    executor::disjunction<detail::is_instance_of_base_impl<Executor, Type>...>;
```
wasn't this supposed to have the word `_any` to the name because of the disjunction?

shrijitsingh99 replied:
Fixed with renaming to `InstanceOfAny`

SergioRAgostinho commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,118 @@
+                                   Executor, executor::cuda_executor> = 0>
+  static bool check(Executor& exec) {
+    return true;
+  }
```
This feels like it should follow the same disjuntion principle i.e., the check should return true of it's an executor of any of the following types `<inline_executor, sse_executor,...>`

shrijitsingh99 replied:
They are separate so as to have different runtime checks for different types of executors.
YOu can also inherit from `executor_runtime_checks` and create custom checks for any one type of executor.

shrijitsingh99 committed:
Switch to shape_type for max threads

SergioRAgostinho commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,118 @@
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
+                "Runtime checks should inherit from executor_runtime_checks");
+  auto num_supported_execs = std::tuple_size<decltype(supported_execs)>::value;
+  auto exec_is_supported = [&](auto& best_fit_exec) {
+    if (!RuntimeChecks::availability_check(best_fit_exec)) return false;
+    auto count = 0;
+    for_each_tuple_until(supported_execs, [&](auto& supported_exec) {
+      if (executor::is_same_template<decltype(best_fit_exec),
+                                     decltype(supported_exec)>::value &&
+          RuntimeChecks::check(supported_exec)) {
+        f(supported_exec);
+        return true;
+      }
+      ++count;
+      return false;
+    });
```
What the difference between `enable_exec_with_priority` and `enable_exec_on_desc_priority`. It  seems to be a matter of order on how executors are picked? The former follows the natural order from `supported_execs` that latter follows the order of `best_fit_executors`. 

We might be able to reduce some of the code duplication between both methods.

shrijitsingh99 replied:
So the idea behind this is that both the functions `enable_exec_with_priority` and `enable_exec_on_desc_priority` choose from one of the executors passed into the function.

The difference lies that `enable_exec_with_priority` checks whether the executors should be chosen in the order the user passed the executors.

`enable_exec_on_desc_priority` checks which executor to be chosen in the order specified by the library i.e. omp > see > inline.

I tried to combine the common functionality but was not able to do so, since we can't modify tuples so re-ordering is not possible of the passed in executors.

SergioRAgostinho replied:
>  since we can't modify tuples so re-ordering is not possible of the passed in executors.

This is how I'm seeing it happen (pseudo-code)

```cpp
auto enable_exec_with_priority(sequence_of_executors_to_try)
{
   // tests the order specified in the sequence
  [...]
}

auto enable_exec_on_desc_priority()
{
    //create tuple with default order established by the library
    sequence_of_executors_to_try = std::tuple<...,...,...>;
    enable_exec_with_priority(sequence_of_executors_to_try);
}
```

is something like this not possible?

shrijitsingh99 replied:
One point missing was, that `enable_exec_on_desc_priority` also requires to pass the executors to try as the sequence of executors passed tells us which executors are supported by the function we are trying to call i.e. which executor types have overload available for them.
```cpp
auto enable_exec_on_desc_priority(sequence_of_executors_to_try)
```

So, what we want is this:
```cpp
auto enable_exec_on_desc_priority(sequence_of_executors_to_try)
{
    //create tuple by rearranging sequence_of_executors_to_try into the order specified by the library
    new_sequence_of_executors_to_try = std::tuple<...,...,...>;
    enable_exec_with_priority(new_sequence_of_executors_to_try);
}
```

I have not figured out a way to do this till now.

SergioRAgostinho replied:
> One point missing was, that `enable_exec_on_desc_priority` also requires to pass the executors to try as the sequence of executors passed tells us which executors are supported by the function we are trying to call i.e. which executor types have overload available for them.

Why is this required? Do we not have a mechanism which allows querying if a given method has a specific executor implementation?

shrijitsingh99 replied:
>  Do we not have a mechanism which allows querying if a given method has a specific executor implementation?

No, I couldn't get this to work.

SergioRAgostinho replied:
We should chat about this at some point then. It's an important feature to have but likely not before GSoC is done.

shrijitsingh99 replied:
Yeah, tracking it in #23

SergioRAgostinho commented on executor/include/executor/trait/common_traits.hpp:
```diff
@@ -38,4 +55,84 @@ struct equality_comparable<
+          typename = typename std::enable_if<
+              I != std::tuple_size<typename std::remove_reference<
+                       TupleType>::type>::value>::type>
+void for_each_tuple_until(TupleType&& t, FunctionType f,
```
`for_each_until_true`?

SergioRAgostinho commented on executor/include/executor/trait/common_traits.hpp:
```diff
@@ -38,4 +55,84 @@ struct equality_comparable<
+  bool exit = f(std::get<I>(std::forward<TupleType>(t)));
+
+  if (exit)
+    for_each_tuple_until(
+        std::forward<TupleType>(t), f,
+        std::integral_constant<size_t,
+                               std::tuple_size<typename std::remove_reference<
+                                   TupleType>::type>::value>());
```
why are you calling a no-op function instead of returning ?

shrijitsingh99 committed:
Rename function to iterate tuple and remove redundant no-op call

shrijitsingh99 committed:
Change shape to be passed as a const reference

shrijitsingh99 committed:
Add support for skipping instantiation of executor function if executor is unavailable

shrijitsingh99 committed:
Rename instance_of_base to InstanceOfAny

shrijitsingh99 committed:
Fix header include

shrijitsingh99 committed:
Add comments

shrijitsingh99 committed:
Add aliases for executors with default properties

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -24,16 +24,15 @@ using namespace executor;
 
 template <typename Executor,
-          typename executor::instance_of_base<inline_executor, Executor> = 0>
+          typename executor::InstanceOfAny<Executor, inline_executor> = 0>
```
Do we have a `InstanceOf` we can use here? since you're testing against a single type.

shrijitsingh99 replied:
Added

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -51,8 +50,18 @@ void mmul(const Executor ex, const MatrixXd& a, const MatrixXd& b,
+  auto mul = [&](auto& exec) { mmul(exec, a, b, c); };
+
+  if (custom_priority)
+    enable_exec_with_priority(mul, inline_executor<>{}, omp_executor<>{});
+  else
+    enable_exec_on_desc_priority(mul, inline_executor<>{}, omp_executor<>{});
```
If you look at these two calls without context it's hard to understand what's the difference between both. Why does the user need to specify executors on `enable_exec_on_desc_priority`. Isn't this one supposed to be a default sequence set by the library?

shrijitsingh99 replied:
Okay so the main purpose the user needs to specify the executor is that there is no way for `enable_exec_on_desc_priority` to know whether the function you passed has an overload for the executor in the default sequence.

Example:
If in `FunctorFilter` there exits an overload for `applyFilter` for OMP and Inline. 
If we don't pass the set of executors to `enable_exec_on_desc_priority` then it will try out SSE executor too since it is specified in the default set of executors by the library, this leads to a compile error.

SergioRAgostinho commented on executor/include/executor/trait/common_traits.hpp:
```diff
@@ -38,4 +55,78 @@ struct equality_comparable<
+struct is_instance_of_any_impl : std::false_type {};
+
+template <template <typename...> class Type, typename... Args>
+struct is_instance_of_any_impl<Type<Args...>, Type> : std::true_type {};
```
`is_instance_of`?

shrijitsingh99 replied:
Not quite sure I understood what you meant.

SergioRAgostinho commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,128 @@
+
+namespace detail {
+
+template <typename Function, typename Executor, typename = void>
+bool execute(Function&& f, Executor& exec) {
+  return false;
+}
+
+template <typename Function, template <typename...> class Executor, typename... Properties,
+          typename std::enable_if<executor::is_executor_available_v<Executor>,
+                                  int>::type = 0>
+bool execute(Function&& f, Executor<Properties...>& exec) {
+  f(exec);
+  return true;
+}
```
I'm surprised you're not getting ambiguous function call errors here. I'm not seeing you disabling the first overload in case the executor is available.

shrijitsingh99 replied:
The second one is more specialized than the first, so if the second is valid the compiler should choose that right?

SergioRAgostinho commented on executor/src/property/blocking.cpp:
```diff
@@ -6,7 +6,7 @@
  *
  */
-#include <executor/property.h>
+#include <executor/property//blocking.hpp>
```
extra forward slash path.

shrijitsingh99 committed:
Fix double slash

shrijitsingh99 committed:
Add InstanceOf trait

shrijitsingh99 committed:
Fix copyright year by updating to 2020

kunaltyagi commented on application/include/mmul.hpp:
```diff
@@ -24,16 +24,15 @@ using namespace executor;
 
 template <typename Executor,
-          typename executor::instance_of_base<inline_executor, Executor> = 0>
+          typename executor::InstanceOf<Executor, inline_executor> = 0>
```
Instead of `= 0` make it `= true` so it reads better

shrijitsingh99 replied:
I thought of that at some point, but even `= false` would be valid, so its slightly misleading.
`= 0` though slightly unclear is unambiguous and more common to see.

kunaltyagi commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,128 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2014-, Open Perception, Inc.
```
Wrong year. It's the year of corona, how could you forget?

kunaltyagi commented on executor/include/executor/trait/is_executor.hpp:
```diff
@@ -15,34 +15,48 @@
+ *  2. Is a CopyConstructible type
+ *  3. Is a EqualityComparable type
+ *
+ *  This concept was finalized in P1660R0 and merged in the draft P0443R13 in
```
:microphone:-drop

kunaltyagi commented on executor/include/executor/trait/is_executor.hpp:
```diff
@@ -15,34 +15,48 @@
-template <typename Executor>
-constexpr bool is_executor_v = is_executor<Executor>::value;
+template <typename T>
+constexpr bool is_executor_v = is_executor<T>::value;
```
static?

shrijitsingh99 replied:
Made it static const. constexpr not possible with lambda pre C++17 due to it being a non-literal type.

kunaltyagi commented on executor/include/executor/trait/is_executor.hpp:
```diff
@@ -15,34 +15,48 @@
-struct is_executor : std::false_type {};
-
 namespace detail {
 const auto noop = [] {};
```
static constexpr?

shrijitsingh99 replied:
Made it static const. constexpr not possible with lambda pre C++17 due to it being a non-literal type.

kunaltyagi replied:
Put a note on it :smile:

kunaltyagi commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,128 @@
+
+struct executor_runtime_checks {
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::inline_executor> = 0>
```
Can't you put multiple types on the same check?

Also, this should be in detail since this is not what the user is supposed to use but is a default 'value'

shrijitsingh99 replied:
You can, but wouldn't we want different checks per executor type? 
How would have multiple types on a single check be used?

I haven't put it in details, since if a user wants to specify his own checks he can do so by inheriting from `executor_runtime_checks`.

kunaltyagi replied:
The function isn't virtual so the downstream user's concerns can be ignored

shrijitsingh99 replied:
I would want this to be virtual ideally, but sadly templates and virtual don't play well.
I am not sure I understood why you want to combine them

kunaltyagi replied:
Them being combined doesn't affect the downstream user in any adverse manner
Combining reduces code
The code is really the same, so combining makes sense

kunaltyagi commented on executor/include/executor/best_fit.hpp:
```diff
@@ -0,0 +1,128 @@
+          typename... SupportedExecutors>
+void enable_exec_with_priority(
+    Function f, std::tuple<SupportedExecutors...> supported_execs) {
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
```
Why is this required? Why not just accept a functor (or lambda) and call that with the executor?

shrijitsingh99 replied:
I am not quite sure what you are referring to by this?
Do you mean `static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
`?

kunaltyagi replied:
I meant why do we need a struct with a `check` static function, why not any functor? Let the user wrap things in a lambda anyhow (such as [overload idiom](https://dev.to/tmr232/that-overloaded-trick-overloading-lambdas-in-c17))

kunaltyagi reviewed (commented):
13/17 :green_circle:

shrijitsingh99 committed:
Convert to static

shrijitsingh99 committed:
Add is_executor_instance_available to check if specialized types of executor are available

shrijitsingh99 committed:
Add tuple filtering trait

Courtesy of @SergioRAgostinho

shrijitsingh99 committed:
Integrate tuple filtering into best fit

Enables reuse of code from enable_exec_with_priority in enable_exec_on_desc_priority

SergioRAgostinho commented on executor/include/executor/best_fit.hpp:
```diff
@@ -79,7 +88,12 @@ void enable_exec_with_priority(
+
+  // This should not happen, at least one executor should always be selected. So either all runtime checks failed which is
+  // incorrect behaviour or no supported executors were passed which should not be done
+  if (!executor_selected)
+      std::cerr<<"No executor selected. All runtime checks returned false or no executors were passed."<<std::endl;
```
I'm surprised this is not resolved already at compile time.

shrijitsingh99 replied:
Well if you have no run-time checks then it is known at compile time whether or not some executor will be selected but adding runtime checks takes away that.

shrijitsingh99 committed:
Use struct due to bug in GCC 5 regarding nested variable template

https://stackoverflow.com/questions/45607450/gcc5-nested-variable-template-is-not-a-function-template

kunaltyagi commented:
Doxy file with 2k lines....

Are you generating documentation on the CI? Can you allow downloading the artifacts so we can read the documentation instead of reading the source to check for errors

shrijitsingh99 commented:
> Doxy file with 2k lines....

Oh no 😱 . Didn't see that.

> Are you generating documentation on the CI? Can you allow downloading the artifacts so we can read the documentation instead of reading the source to check for errors

Will take a look and it.

shrijitsingh99 committed:
Add CI for Ubuntu 16.04

shrijitsingh99 committed:
Add documentation

shrijitsingh99 committed:
Add member function to set max number of threads for omp executor

shrijitsingh99 committed:
Add environment variable checks to best fit and limit max threads in best fit omp executor

shrijitsingh99 committed:
Add support for derived executors in instance_of_any

shrijitsingh99 committed:
Add can_query trait

shrijitsingh99 committed:
Add documentation

shrijitsingh99 committed:
Rename basic_executor_property to base_executor_property

shrijitsingh99 committed:
Add documentation

shrijitsingh99 committed:
Update documentation formatting

shrijitsingh99 committed:
Add documentation

shrijitsingh99 committed:
Add documentation

kunaltyagi commented on executor/include/executor/trait/is_executor.hpp:
```diff
@@ -38,7 +38,9 @@ struct contains_execute<
 
 /**
- * \brief A given type T is an Executor if it satisfies the following
+ * \brief Checks whether the given type T is an executor
```
You can remove the word `T` and add a `\tparam T details`

shrijitsingh99 replied:
Added. 
Will add to other locations as well

kunaltyagi commented on executor/include/executor/trait/can_require.hpp:
```diff
@@ -17,17 +17,41 @@ namespace executor {
- * by all properties
+ * \brief Checks if the given Executor supports the Property
+ *
+ *\details This is checked through a template variable static_query_v which is
```
space after star

kunaltyagi commented on executor/include/executor/trait/can_prefer.hpp:
```diff
@@ -15,6 +15,31 @@
+ * that property
+ *
+ * If all the above conditions are met, prefer customization point is valid.
+ * If it is possible to call the require customization point, then it is called
```
If it is possible, then the require customization point is called ...

kunaltyagi commented on executor/include/executor/property/base_property.hpp:
```diff
@@ -14,29 +14,86 @@
+ *
+ * \details Currently for the property to be applicable T should be an executor type.
+ * In proposal P1393R0 there are no conditions specified for a property to be applicable
+ * ans is left to the implementer to decide.
```
ans?

shrijitsingh99 replied:
Should have been `and`. 
Fixed.

kunaltyagi commented on executor/include/executor/property/base_property.hpp:
```diff
@@ -14,29 +14,86 @@
+ * without a base class.
+ *
+ * A CRTP design is used as static query needs to be able to access the
+ * DerivedProperty in order to query the overloaded query member function in an
```
`in order to query the overloaded query member function`

Reword this part to not have 2 meanings of query: the English verb and the member function

shrijitsingh99 replied:
`in order to call the overloaded query member`
Better?

shrijitsingh99 committed:
Add doxygen config

shrijitsingh99 committed:
Add doxygen generation to CI

shrijitsingh99 committed:
Update doxygen config

shrijitsingh99 commented:
Added documentation generation & deployment as part of CI.
Documentation can be viewed [here](http://shrijitsingh.com/unified-executors/)

shrijitsingh99 committed:
Fix grammatical mistakes

shrijitsingh99 committed:
Update documentation
